File: 프로그래머스/Lv.2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque 

def solution(maps):
    
    #maps = [[1, 1]]
    dr, dc = [-1,1,0,0], [0,0,-1,1]
    que = deque()
    que.append([0,0])
    N,M=len(maps), len(maps[0])
    

    while que:
        r,c = que.popleft()
        for idx in range(4):
            nr, nc = r+dr[idx], c + dc[idx]
            ## 범위 밖인 경우 OR 벽인 경우
            if not(0<=nr<N) or not(0<=nc<M) or maps[nr][nc]==0: 
                continue
            ## 최초 방문인 경우 
            if maps[nr][nc]==1:
                maps[nr][nc]=maps[r][c]+1
                que.append([nr,nc])
    
    answer = -1 if maps[-1][-1]==1 else maps[-1][-1]
    return answer

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const [N,M] = [maps.length, maps[0].length]
    var answer = -1;
    const dir = [[0,-1],[0,1],[-1,0],[1,0]]
    const que = [[N-1,M-1]]
    while(que.length>0){
        const [r,c] = que.shift()
        if(r==0 && c==0){
            answer=(maps[r][c])
            break
        }
        for(let idx=0;idx<4;idx++){
            const [nr, nc] = [r+dir[idx][0],c+dir[idx][1]]
             // console.log(nr, nc)
            
            /*범위여부*/
            if(nr>=N||nr<0 || nc<0||nc>=M ||maps[nr][nc]!=1){continue}
            
            maps[nr][nc]=maps[r][c]+1
            que.push([nr, nc])
            
        }
    }
    
  
    return answer;
}